### PR TITLE
fix/misc/fix-e2e : Add tests to e2e branch

### PR DESCRIPTION
**Pull request by charliemangano**
# Add tests to e2e branch
**Contrary to usual, this PR is not asking to merge into `main` but into `test/end-to-end`**

## Description
This PR fixes the already written end-to-end tests to have them pass the CI. It closes part of the issue #211.
Please check #211 for the full explanation.

## Changes
I mostly just increased the timeout to 60s, to be sure that even our slow CI waits long enough for the actions to complete.

I also added `.performScrollTo` everywhere applicable to account for the CI's emulator's small screen. The tests also now clear the text fields before performing an input on them. We didn't notice this before because the input was fast enough that the fields hadn't fetched the data from our repository yet, but it does produce errors sometimes, so I changed it.

#### Added
None
#### Modified
- `app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM1.kt`: changed name of test class to `EndToEndM1` for coherence and increased timeout
- `app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt`: increase timeout, added `performScrollTo` calls, cleared fields before trying to input new values and added docs
#### Removed
None
## Dependencies Added
None

## Testing
All pass on local, as before. Now systematically passes on CI. We might still have problems if the CI is especially slow but that shouldn't happen often.

### CODE REVIEW

File: app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt [hunk 1]
```diff
@@ -45,60 +47,129 @@ class EndToEndM2 : TestCase() {
...
     composeTestRule
         .onNodeWithTag(ProfileScreens.NAME_INPUT_FIELD)
+        .performScrollTo()
+        .assertIsDisplayed()
+        .performTextClearance()
+    composeTestRule
+        .onNodeWithTag(ProfileScreens.NAME_INPUT_FIELD)
+        .performScrollTo()
         .assertIsDisplayed()
         .performTextInput(name)
```
Comment (francelu) [inline app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt:108]: Good point to add the scrolls and text clearance.  
However, I wonder if it is necessary to `composeTestRule(Tag).perfromScrollTo().assertIsDisplayed()` twice here with the first a text clearance and the second a text input.
```suggestion
    composeTestRule
        .onNodeWithTag(ProfileScreens.NAME_INPUT_FIELD)
        .performScrollTo()
        .assertIsDisplayed()
        .performTextClearance()
        .performTextInput(name)
```

Same remark for other instances.

File: app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt [hunk 2]
```diff
@@ -45,60 +47,129 @@ class EndToEndM2 : TestCase() {
...
+        .assertTextEquals(name)
     composeTestRule
         .onNodeWithTag(ProfileScreen.DESCRIPTION_FIELD)
+        .performScrollTo()
         .assertExists()
         .assertTextEquals(description)
+    composeTestRule
+        .onNodeWithTag(ProfileScreen.NO_REVIEWS_CARD)
+        .performScrollTo()
+        .assertExists() // TODO: change once implemented the
```
Comment (francelu) [inline app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt:168]: `the` ... ? Statuses I think you meant :)

**Review by francelu — APPROVED** [review]
## Summary
LGTM. It's good to add a large time out. You also ensured a good performance of the tests using the scroll and text clearance.

## Important
#### Code Quality
- Code is readable, consistent, and clear. Good job !

#### Functionality
- Edge cases (notably small emulator) are handled thanks to the scrolls.
- Good to notice and add the text clearance.

#### Performance
- This helps our other branches pass the CI more frequently as the e2e tests won't be bothering at each commit on PRs.
- **NOTE**: your wiki about how we should be regularly testing the e2e is excellent, easy to follow and understand. Lovely job !

#### Security
- Unchanged.

## Testing
- Good adaptation and overall consistency in nomenclature and others.

## Documentation
- You adapted documentation as necessary. Good !

## Suggestions
- Check my comments :)
- Else none.

## Steps before PR **merged**
- Reply to my questions and propositions.

File: app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt (hunk 1 above)
Comment (charliemangano) [reply app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt:108]: I wondered as well, but it turns out `performTextClearance` returns a `Unit` and not the `SemanticsNodeInteraction` we would need to chain these calls :(

File: app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt (hunk 2 above)
Comment (charliemangano) [reply app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt:168]: Ah yes, fixing that!

File: app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt [hunk 3]
```diff
@@ -24,7 +26,7 @@ import org.junit.Test
 import org.junit.runner.RunWith
 
 private const val TAG = "EndToEnd2"
-private const val TIMEOUT = 10000L // 10 seconds, adjust for slower devices, networks and CI
+private const val TIMEOUT = 50000L // 50 seconds, adjust for slower devices, networks and CI
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt:29]: It is kinda arbitrary, but still a bit informed: @agonzalez-r tried before with 10s, 20s and I think 30s and that did not help the CI :( I myself created a secondary branch where I tried 10, 20, 30, 40 and 50, and 50s was the one that passed (almost) every time. I did not try above that though.

File: app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt (hunk 1 above)
Comment (francelu) [reply app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt:108]: Too bad then... :(

File: app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt:29]: Okay, fair enough. I would round it up to a minute (why not ? :) ).

File: app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt [hunk 4]
```diff
@@ -45,60 +47,129 @@ class EndToEndM2 : TestCase() {
...
     }
 
-    // Profile Screen, check if the changes are saved
+    // Profile Screen
+    // User arrives on Profile Screen and see their new data displayed
     composeTestRule.waitForIdle()
+    composeTestRule.waitUntil(
+        TIMEOUT) { // need to wait because very first recomposition has not fetched data yet
+          try { // trick waitUntil into thinking this counts as a SemanticNodeInteraction
+            composeTestRule
```
Comment (agonzalez-r) [inline app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt:144]: I don't understand why you encapsulate the name assertion, when 
```
composeTestRule.waitUntil(TIMEOUT) {
      composeTestRule.onAllNodesWithTag(ProfileScreen.SCREEN).fetchSemanticsNodes().size == 1
    }
```
which is called just before, should already ensure that the profile screen is fully loaded and all the components are displayed?

Comment (charliemangano) [reply app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt:144]: There are two different things we are waiting for: 
1. for the navigation to the next screen to complete: we wait until the screen is asserted as existent
2. for the data to be fetched from the repository: we wait until the correct data is displayed in at least on of the fields

If we don't wait for 2., the tests take the string displayed at first composition ("Error loading X, try again later") instead of the (correctly but slowly) fetched data.

**Review by agonzalez-r — APPROVED** [review]
LGTM! 
The tests pass locally and modifications directly addresses previous issues with the CI.

File: app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/androidTest/java/com/android/periodpals/endtoend/EndToEndM2.kt]: Done!
